### ISSUE TITLE
Select output denominations based on mint statistics

### ISSUE THREAD

**Issue by elsirion** [fedimint/minimint#116]
To avoid issuing denominations with low anonymity set there should be a federation API endpoint that shows issuances per denomination. The client should use that information to choose output denominations. 

Random thought: the denominations should be uniformly distributed for low amounts that everyone uses. Larger amounts would see less usage but possibly due to this fix none instead. So we should see the highest  actually used denomination have the highest usage because all the theoretically higher ones would be split into this one. We could use the amplitude of that behavior to judge when to bootstrap a higher denomination.

**Comment by MaxHillebrand** [comment]
Very interesting, but could an evil federation mislead users with pumping up those statistics, claiming fake usage for a perceived anonset?

**Comment by elsirion** [comment]
Yes, in a sense it degrades information theoretical privacy to computational privacy under the assumption that not more than f federation members are malicious (you can just ask multiple ones to cross-check or make the federation threshold-sign the statistic). Why I think this is ok:

Information theoretical privacy only holds under some not fully realistic assumptions (everyone having the same probability distribution of owning tokens of certain denominations). Given this fact my proposal is a strict improvement, the worst the federation can do is downgrading the user's privacy to the current state of affairs. They can also make you not use denominations that would be ok to use, but that actually helps privacy but reduces scalability. Might be used to collect more fees?

Lastly, for the community banking/social recovery aspect we'll need to make this trade-off anyway to trust the federation for privacy too (they could conspire to decrypt your recovery seed without you asking for it).

**Comment by jkitman** [comment]
@MaxHillebrand we could use your research on denominations in WabiSabi 2.0 to help us determine the best denominations to use for our ecash to maximize privacy without creating too many coins to sign.  I couldn't find much in the whitepaper, is there anything you could point us to?

**Comment by MaxHillebrand** [comment]
We did some experiments here github.com/nopara73/sake, it's very coinjoin specific, but these standard denominations might already be good for you.

Powers of 2
Powers of 3
Powers of 3 times 2
Powers of 10
Powers of 10 times 2
Powers of 10 times 5

We then removed those denominations that are too close to each other.

iirc, we have 42 denominations between 5000 sats and 1 btc.

Here a short code snippet to generate them https://gist.github.com/lontivero/ac4d86d060ae468f703e60bd62c12b42 

Here the Wasabi code for generating the denominations https://github.com/zkSNACKs/WalletWasabi/blob/master/WalletWasabi/WabiSabi/Client/AmountDecomposer.cs 

For theoretical research a good place to start is Hamming weight https://en.wikipedia.org/wiki/Hamming_weight 


Let me know how I can help further, we could jump on a call.

**Comment by jkitman** [comment]
@MaxHillebrand thanks for those links, I did a bit more research and here are my thoughts on the denominations in fedimint:

When choosing denominations there is always a trade-off between the size of the 
anonymity set and the number of notes created.  Naively, the best anonymity set would be to always use 1msat notes, but the computation/storage costs make it infeasible.

Computing the optimal denomination values to minimize the amount of notes given a fixed number of denominations and upper-bound on spending amounts is computationally hard.  Furthermore that would assume a uniform distribution of amounts which isn’t true in practice.  A quote from the literature:

>One approach reduces the problem to determining the smallest set of denominations capable of handling a given range of transactions.  Analogizing to Bachat’s problem of establishing a standard
set of weights, these authors typically characterize the optimal structure of denominations
as a powers-of-three.
>
>Another approach argues the optimal denomination structure would minimize the number of coins and banknotes used in the average transaction (e.g., Van Hove and Heyndels 1996; Van Hove 2001). Following the “principle of least effort,” these authors claim the optimal structure of denominations is the powers-of-two.


For our purposes, I think choosing either would be fine, so long as we include both a 1msat and 1sat denomination for making LN or on-chain change.  There isn’t any literature I could find on how denominations would impact anonymity sets, but I think so long as we choose a reasonably small number of denominations we can solve for privacy in other ways.

For instance, we can assume that the number of notes of each denomination are publicly known and associated with specific transactions, since we allow clients to download the epoch history and under our threat model we assume that there may be malicious federation members who seek to deanonymize users.

When spending ecash, users could check the total number of notes that are outstanding of the denominations they are spending.  If the anonymity set isn’t high enough, they could reissue different denominations of ecash first (perhaps periodically this is done as well to avoid timing analysis).

Furthermore, I think when a fedimint is created it is important that the guardians seed the federation with some bitcoin.  This could help build an anonymity set of different denominations and also is necessary so the federation can support simultaneous on-chain peg-outs.

**Comment by MaxHillebrand** [comment]
Yes, I think in your case less denominations is preferable as bandwidth and storage does not seems like a big bottleneck. I'd guess powers of two or three are "good enough". 

This is also something that can be experimented with in different federations.

A long term solution might be to research federated anonymous amount eCash tokens. WabiSabi hides the amount from a single central issuer. Maybe you can again find some fancy threshold crypto magic?
